aeppl/truncation.py

```diff
@@ -0,0 +1,126 @@
+    logcdf = _logcdf(op.base_op, value, *rv_params, name=name, **kwargs)
+    if op.base_op.name:
+        logprob.name = f"{op.base_op.name}_logprob"
+        logcdf.name = f"{op.base_op.name}_logcdf"
+
+    is_lower_bounded, is_upper_bounded = False, False
+    if not (
+        isinstance(upper_bound, TensorConstant) and np.all(upper_bound.value == np.inf)
+    ):
+        is_upper_bounded = True
+        logprob = at.switch(
+            at.eq(value, upper_bound),
+            at.log(1 - at.exp(logcdf)),
+            at.switch(at.gt(value, upper_bound), -np.inf, logprob),
+        )
+    if not (
+        isinstance(lower_bound, TensorConstant) and np.all(lower_bound.value == -np.inf)
+    ):
+        is_lower_bounded = True
+        logprob = at.switch(
+            at.eq(value, lower_bound),
+            logcdf,
+            at.switch(at.lt(value, lower_bound), -np.inf, logprob),
+        )
+
+    if is_lower_bounded and is_upper_bounded:
+        logprob = Assert("lower_bound <= upper_bound")(
+            logprob, at.all(at.le(lower_bound, upper_bound))
+        )
+
+    return logprob
+
+
+@local_optimizer(tracks=[Elemwise])
+def censor_rvs(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
+    # TODO: Canonicalize x[x>ub] = ub -> clip(x, x, ub)
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+    if rv_map_feature is None:
+        return
+
+    if not isinstance(node.op, Elemwise) or not isinstance(node.op.scalar_op, Clip):
+        return
+
+    clipped_var = node.outputs[0]
+    if clipped_var not in rv_map_feature.rv_values:
+        return
+
+    base_var, lower_bound, upper_bound = node.inputs
+
+    if base_var in rv_map_feature.rv_values:
+        warnings.warn(
+            f"Value variables were assigned to both the input ({base_var}) and "
+            f"output ({clipped_var}) of a censored random variable."
+        )
+        return
+
+    if base_var.owner is None or not isinstance(base_var.owner.op, RandomVariable):
+        return
+
+    base_op = base_var.owner.op
+
+    censored_rv = CensoredRV(
+        "censored",
+        base_op.ndim_supp,
+        list(base_op.ndims_params) + [base_op.ndim_supp] * 2,
+        base_op.dtype,
+        inplace=False,
+        base_op=base_op,
+    )
+
+    # Replace bounds by +-inf if `y = clip(x, x, ?)` or `y=clip(x, ?, x)`
+    # This is used in `censor_logprob` to generate a more succint logprob graph
+    # for one-sided censored random variables
+    # TODO: This will probably fail for multivariate variables
+    lower_bound = lower_bound if (lower_bound is not base_var) else -np.inf
+    upper_bound = upper_bound if (upper_bound is not base_var) else np.inf
+
+    censored_node = censored_rv.make_node(
+        *base_var.owner.inputs,
+        lower_bound,
+        upper_bound,
+    )
+
+    censored_node_out = censored_node.outputs[1]
+
+    if not censored_node_out.type == clipped_var.type:
+        return
```

Comment:
This would make more sense as an `assert`.  Even so, this shouldn't be a necessary check, because the logic that determines the output type is actually in `RandomVariable` and is specified directly in the above `CensoredRV` constructor call.

NB: You might be able to simplify the use of `CensoredRV` by extracting the `dtype`&mdash;and other poperties from `base_op`&mdash;in `CensoredRV.__init__`.